Documentation/proposals/202405-scrapeconfig-graduation.md

```diff
@@ -0,0 +1,130 @@
+
+1. **Add New SD Configurations** (Including and not limited to):
+   - **`dockerswarm_sd_config`**: Docker Swarm is a popular container orchestration tool. Supporting Docker Swarm SD will cater to users leveraging Docker Swarm for their deployments.
+   - **`puppetdb_sd_config`**: PuppetDB is a storage service that keeps data generated by Puppet. Integrating PuppetDB SD will be beneficial for users who utilize Puppet for configuration management.
+   - **`lightsail_sd_config`**: Amazon Lightsail is a simplified cloud platform. Including Lightsail SD will assist users who prefer Lightsail for its ease of use and cost-effectiveness.
+
+It has been noted that it might not be necessary to cover all of the Service discoveries since some of them are very niche and rarely used. But it would not hurt to cover all of them eventually.
+Newer SDs can be added as part of feature requests, which we hope to prompt users to do if the SD they use is not yet supported.
+
+#### Filling Gaps in The CRD
+
+Ensuring that the ScrapeConfig CRD accurately reflects Prometheus's `scrape_config` is essential for maintaining consistency and completeness. This involves including all relevant fields and ensuring they are correctly implemented. Our approach includes:
+
+1. **Include Missing Fields**:
+   - We will incorporate all fields present in Prometheus `scrape_config` into the ScrapeConfig CRD. However, there are specific exceptions:
+     - **`job_name`**: To prevent issues related to non-unique job names, we are introducing an equivalent `JobName` field, which ensures safer and more predictable configurations.
+     - **`body_size_limit`**: This field is managed globally through `CommonPrometheusFields`. Additionally, users can set `body_size_limit` at the ScrapeConfig CRD level for specific scrape jobs, providing flexibility without redundancy.
+   - Include missing fields such as:
+     - `scrape_classic_histograms`
+     - `follow_redirects`
+     - `enable_http`
+     - `oauth2`
+     - `native_histogram_bucket_limit`
+     - `native_histogram_min_bucket_factor`
+
+2. **Review and Standardize**:
+   - **Comprehensive Review**: We will conduct a thorough review of all existing SD configurations to ensure they meet the required standards. This includes verifying the presence of all necessary fields and checking for consistency in naming conventions.
+   - **Standardization**: By standardizing naming conventions and field presence across both existing and new SD configurations, we ensure that the ScrapeConfig CRD is intuitive and easy to use for developers and users alike.
+   - **Documentation and Examples**: We will update all related documentation and provide clear examples to reflect these changes. This will aid users in understanding and utilizing the enhanced SD configurations effectively.
```

Comment:
Nice to know that. Let me know for documentation enhancement. It would also help me to keep adding missing features in the documentation asap.

Reply:
For documentation I have some examples in https://github.com/slashpai/prometheus-operator-examples/tree/main/scrape_config which can be used